cpp/src/arrow/filesystem/localfs.cc

```
@@ -309,6 +314,208 @@ Result<std::vector<FileInfo>> LocalFileSystem::GetFileInfo(const FileSelector& s
+    ARROW_ASSIGN_OR_RAISE(
+        auto part_gen, DiscoverDirectoryFiles(std::move(selector), std::move(fs_opts)));
+    return dir_readahead > 1
+               ? MakeSequencedMergedGenerator(std::move(part_gen), dir_readahead)
+               : MakeConcatenatedGenerator(std::move(part_gen));
+  }
+
+ private:
+  /// The class, which implements iterator interface to traverse a given
+  /// directory at the fixed nesting depth, and possibly recurses into
+  /// sub-directories (if specified by the selector), spawning more
+  /// `DiscoveryImplIterators`, which feed their data into a single producer.
+  class DiscoveryImplIterator {
+    const PlatformFilename dir_fn_;
+    const int32_t nesting_depth_;
+    const FileSelector selector_;
+    const uint32_t batch_size_;
+
+    FileInfoGeneratorProducer file_gen_producer_;
+    FileInfoVector current_chunk_;
+    std::vector<PlatformFilename> child_fns_;
+    size_t idx_ = 0;
+    bool initialized_ = false;
+
+   public:
+    DiscoveryImplIterator(PlatformFilename dir_fn, int32_t nesting_depth,
+                          FileSelector selector,
+                          FileInfoGeneratorProducer file_gen_producer,
+                          uint32_t batch_size)
+        : dir_fn_(std::move(dir_fn)),
+          nesting_depth_(nesting_depth),
+          selector_(std::move(selector)),
+          batch_size_(batch_size),
+          file_gen_producer_(std::move(file_gen_producer)) {}
+
+    /// Pre-initialize the iterator by listing directory contents and caching
+    /// in the current instance.
+    Status Initialize() {
+      auto result = arrow::internal::ListDir(dir_fn_);
+      if (!result.ok()) {
+        auto status = result.status();
+        if (selector_.allow_not_found && status.IsIOError()) {
+          auto exists = FileExists(dir_fn_);
+          if (exists.ok() && !*exists) {
+            return Status::OK();
+          } else {
+            return exists.ok() ? arrow::Status::UnknownError(
```

If `exists` is ok you should simply propagate the error not-found error, not create another one.
This would make the final code look like this probably:
```c++
      if (!result.ok()) {
        if (selector_.allow_not_found && status.IsIOError()) {
          ARROW_ASSIGN_OR_RAISE(bool exists, FileExists(dir_fn));
          if (!exists) {
            return Status::OK();
          }
        }
        return status;
      }
```

(incidentally, this is a similar snippet as in `StatSelector()`)